DllImportGenerator/designs/Pipeline.md

```diff
@@ -0,0 +1,47 @@
+# P/Invoke Generation Pipeline
+
+The P/Invoke source generator is responsible for finding all methods marked with `GeneratedDllImportAttribute` and generating code for their implementations (stubs) and corresponding P/Invokes that will be called by the stubs. For every method, the steps are:
+
+1. [Process the symbols and metadata](#symbols-and-metadata-processing) for the method, its parameters, and its return type.
+1. [Determine the marshalling generators](#marshalling-generators) that will be responsible for generating the stub code for each parameter and return
+1. [Generate the stub code](#stub-code-generation) and corresponding P/Invoke
+1. Add the generated source to the compilation.
+
+The pipeline uses the Roslyn [Syntax APIs](https://docs.microsoft.com/dotnet/api/microsoft.codeanalysis.csharp.syntax) to create the generated code. This imposes some structure for the marshalling generators and allows for easier inspection or modification (if desired) of the generated code.
+
+## Symbol and metadata processing
+
+The generator processes the method's `GeneratedDllImportAttribute` data, the method's parameter and return types, and the metadata on them ([`LCIDConversionAttribute`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.lcidconversionattribute), [`MarshalAsAttribute`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshalasattribute), [struct marshalling attributes](StructMarshalling.md)). This information is used to determine the corresponding native type for each managed parameter/return and how they will be marshalled.
+
+## Marshalling generators
+
+Each parameter and return for the method is handled by an [`IMarshallingGenerator`](../DllImportGenerator/Marshalling/MarshallingGenerator.cs) instance. The processed information for each parameter and return type is used to determine the appropriate marshalling generator for handling that type. Support for different types can be added in the form of new implementations of `IMarshallingGenerator`.
+
+The marshalling generators are responsible for generating the code for each [stage](#stages) of the stub. They are intended to be stateless, such that they are given all the data ([`TypePositionInfo`](../DllImportGenerator/TypePositionInfo.cs)) for which they need to generate marshalling code and the context ([`StubCodeContext`](../DllImportGenerator/StubCodeContext.cs)) under which that code should be generated.
+
+## Stub code generation
+
+Generation of the stub code happens in stages. The marshalling generator for each parameter and return is called to generate code for each stage of the stub. The statements and syntax provided by each marshalling generator for each stage combine to form the full stub implementation.
+
+### Stages
```

Comment:
The stages look good. One question is regarding positional and implicit arguments. For example LCID creates an argument to pass and `PreserveSig` adds an out param to the method signature. Is this scheme able to handle that efficiently? Do we have an example/pattern for how handle those cases?

Reply:
I was thinking most of the handling would be on creating all the `TypePositionInfo`s during processing of symbols/metadata - setting `Managed/UnmanagedIndex` properties accordingly and creating ones to represent an LCID argument / out param / HResult if needed. We could have specific marshalling generators for the LCID argument and HResult checking.

Then, going through the stages should not be affected other than a bit of extra cruft around declaring variables for the return value of the stub and the invoke (if needed) and determining identifier names.

I didn't want to deal with actually handling those cases yet. I can add some placeholder TODOs to indicate what I had in mind.

Reply:
> LCID argument 

FWIW, LCID arguments is an obscure legacy feature. I would be perfectly happy to not support it for the source generators if it were to add even a bit of extra complexity anywhere.